### ISSUE TITLE
Sigill When encoding

### ISSUE THREAD

**Issue by 1480c1** [OpenVisualCloud/SVT-HEVC#274]
With ubuntu 19.04 on Xeon X3430, when trying to encode, the program reports
``` none
Program received signal SIGILL, Illegal instruction.
0x000055555556daea in CheckXcr0Zmm () at /code/SVT-HEVC/Source/Lib/Codec/EbEncHandle.c:215
215         __asm__ ("xgetbv" : "=a" (xcr0) : "c" (0) : "%edx" );
```

SVT-AV1 is fine, only SVT-HEVC doesn't work. Haven't tested SVT-VP9 yet

**Comment by 1480c1** [comment]
I do think that it's probably because of the cpu, but I wonder why SVT-AV1 works fine

**Comment by tianjunwork** [comment]
Here is the cpu arch of your X3430:  https://en.wikipedia.org/wiki/Lynnfield_(microprocessor)
AV1 compiles with -mavx. HEVC doesn't specify any. Native is disabled by #243.
I am not sure why -mavx works fine with your X3430.

**Comment by 1480c1** [comment]
For the cpu flags part, the c compilers (gcc and clang, idk about others) do not care about if the build computer's cpu can support certain instruction sets unless `-march=native` is set (still doesn't override `-mavx2` etc unless `-no-mavx2` is set) so the gcc on that computer can use `-mavx` or `-mavx2` and generate avx and avx2 code, even though it cannot run it.
Main benefit is for compiling for non-native hosts (Older server with multiple cores generate code for a faster newer cpu that might have less cores but newer instruction sets)

TLDR: The build cpu doesn't affect if gcc can accept `-mavx` or others

**Comment by tianjunwork** [comment]
So you didn't compile code on X3430? Just run the encoder app on it?  Where the app is compiled?

**Comment by 1480c1** [comment]
I did compile on x3430, it's just that GCC doesn't care and will compile avx and avx2 instructions

**Comment by tianjunwork** [comment]
I think you have to specifically use -mavx to make SVT-HEVC work on X3430 which support up to SSEx.

**Comment by 1480c1** [comment]
The cpu doesn't support avx, the highest it supports is sse4.1

**Comment by 1480c1** [comment]
If I use `-mavx`, GCC will convert sse functions to avx

**Comment by tianjunwork** [comment]
I doubt sse is converted to avx. Since we don't have avx code, use -mavx will fall back to SSEx. I guess that's why AV1 works. Let me confirm with other folks.

**Comment by 1480c1** [comment]
``` none
GCC depresses SSEx instructions when -mavx is used. Instead, it generates new AVX instructions or AVX equivalence for all SSEx instructions when needed.
```
https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#x86-Options

**Comment by tianjunwork** [comment]
yes, you are right, gcc may compile SSEx to AVX. 
AV1 has some extra SSEx code which behaves differently with HEVC.

**Comment by Austin-Hu** [comment]
Suggest to merge the PR #376, which is a reasonable fix addressing the illegal instruction problem with xgetbv.

@1480c1 , to fully fix this issue, please refer to the [commit](https://github.com/Austin-Hu/SVT-HEVC/commit/d708ea7066684ee17d9155d312e90b6d695a9b5f) to confirm that the "Illegal instruction" issue wouldn't happen in IntraModeVerticalChroma_AVX2_INTRIN(), but it occurred again in another following function IntraModeDCChroma_AVX2_INTRIN(). And we would set it as low priority and resolve this issue in the future. Thanks!

**Comment by Austin-Hu** [comment]
BTW, we'd remember the comment thread in the closed PR #362 , to understand the root cause of this issue.

**Comment by 1480c1** [comment]
Yes, With your commit, rebased on master, sigill is no longer raised on IntraModeVerticalChroma_AVX2_INTRIN, but on IntraModeDCChroma_AVX2_INTRIN instead. This definitely can be marked as low priority since most people using SVT-HEVC are most likely going to be using a more up to date cpu

**Comment by Austin-Hu** [comment]
> Yes, With your commit, rebased on master, sigill is no longer raised on IntraModeVerticalChroma_AVX2_INTRIN, but on IntraModeDCChroma_AVX2_INTRIN instead. This definitely can be marked as low priority since most people using SVT-HEVC are most likely going to be using a more up to date cpu

Thanks for your confirmation, @1480c1 ! So let's keep the issue open for monitoring.

**Comment by 1480c1** [comment]
It seems the reason why was that although `GetCpuAsmType()` was returning the correct information, some of the arrays were using `ASM_TYPES & PREAVX2_MASK` which was evaluating to 1 on non-avx systems, which was in turn using the AVX2 functions in the arrays instead of the C functions

**Comment by tianjunwork** [comment]
Addressed by #376. Close this issue.
